src/CodeGeneration.Roslyn.Engine/CompilationGenerator.cs

```diff
@@ -74,23 +92,18 @@ public class CompilationGenerator
         public IEnumerable<string> EmptyGeneratedFiles => this.emptyGeneratedFiles;
 
         /// <summary>
-        /// Gets or sets the directory with the project file.
+        /// Gets the directory with the project file.
         /// </summary>
-        public string ProjectDirectory { get; set; }
+        public string ProjectDirectory { get; }
 
         /// <summary>
         /// Runs the code generation as configured using this instance's properties.
         /// </summary>
         /// <param name="progress">Optional handler of diagnostics provided by code generator.</param>
         /// <param name="cancellationToken">Cancellation token to interrupt async operations.</param>
         /// <returns>A <see cref="Task.CompletedTask"/>.</returns>
-        public async Task GenerateAsync(IProgress<Diagnostic> progress = null, CancellationToken cancellationToken = default)
+        public async Task GenerateAsync(IProgress<Diagnostic> progress, CancellationToken cancellationToken = default)
         {
-            Verify.Operation(this.Compile != null, $"{nameof(Compile)} must be set first.");
-            Verify.Operation(this.ReferencePath != null, $"{nameof(ReferencePath)} must be set first.");
-            Verify.Operation(this.PluginPaths != null, $"{nameof(PluginPaths)} must be set first.");
-            Verify.Operation(this.IntermediateOutputDirectory != null, $"{nameof(IntermediateOutputDirectory)} must be set first.");
```

Comment:
Public API behavior (null validation) shouldn't be removed. NRTs don't change that - they don't actually prevent null from being passed.

Then again, this whole class had it's shape changed in this PR; @AArnott what's your PoV here?

Reply:
> Public API behavior (null validation) shouldn't be removed. NRTs don't change that - they don't actually prevent null from being passed.

That's true. I will add them

Reply:
But the caller can't screw this up. This isn't argument validation but internal state validation, which the class now is (supposedly) guaranteeing is already correct. So I don't mind removal of this, *if* indeed the arg validation is added to the public constructor.

Reply:
Seems to have been added in ctor.